### PR TITLE
Various refactorings

### PR DESCRIPTION
**Pull request by marcelm**
Here are some refactorings of the alignment code. I’m doing this as part of #26 "cleaning up as I go" while trying to understand how the alignment/mapping is done.

* Change `HammingToCigarOQX2` so that it returns an `aln_info` struct to make it more compatible with `ssw_align` (also rename it to `hamming_align`)
* Move all NAM-related stuff to `nam.cpp/.h` (this was very helpful as it made it clear that `hit` and `Hit` are only used privately by the nam-finding functions)
* Move all PAF-output-related stuff to `paf.cpp/.h`
* Convert to 1-based coordinates at the last possible moment when writing the SAM output. This way, the details of the file format do not leak into the alignment code. This makes it simpler, for example, to switch from SAM to BAM output
if we wanted to because BAM uses 0-based coordinates.

With the above, `aln.cpp` went from 2340 to 1900 lines.

### CODE REVIEW

File: src/aln.cpp [hunk 1]
```diff
@@ -27,7 +27,7 @@ static inline bool score(const nam &a, const nam &b) {
     return a.score > b.score;
 }
 
-inline aln_info ssw_align(const std::string &ref, const std::string &query, int match_score, int mismatch_penalty, int gap_opening_penalty, int gap_extending_penalty) {
+inline aln_info ssw_align(const std::string &ref, const std::string &query, alignment_params parameters) {
```
Comment (ksahlin) [inline src/aln.cpp:19]: Is it nicer/faster to send alignment_params by reference?

**Comment by ksahlin** [comment]
This looks great!

File: src/aln.cpp (hunk 1 above)
Comment (marcelm) [reply src/aln.cpp:19]: I thought a bit about this when making that change before I settled on this variant. `alignment_params` consists of four 32-bit fields, so they can be passed using two 64-bit registers. When they are passed by reference, only one 64-bit register is needed for the memory address. On the other hand, then a memory access is needed within the function, which is slower than accessing a register. In the end, I am sure this does not make a difference at all and both variants are fine.

Comment (ksahlin) [reply src/aln.cpp:19]: thanks for the explanation!

**Pull request merged**